rust/main/utils/run-locally/src/main.rs

```diff
@@ -536,13 +474,151 @@ fn main() -> ExitCode {
+        .hyp_env("CHAINS_TEST1_RPCCONSENSUSTYPE", "fallback")
+        .hyp_env(
+            "CHAINS_TEST2_CONNECTION_URLS",
+            "http://127.0.0.1:8545,http://127.0.0.1:8545,http://127.0.0.1:8545",
+        )
+        .hyp_env(
+            "CHAINS_TEST1_BATCHCONTRACTADDRESS",
+            multicall_address_string.clone(),
+        )
+        .hyp_env("CHAINS_TEST1_MAXBATCHSIZE", "5")
+        // by setting this as a quorum provider we will cause nonce errors when delivering to test2
+        // because the message will be sent to the node 3 times.
+        .hyp_env("CHAINS_TEST2_RPCCONSENSUSTYPE", "quorum")
+        .hyp_env(
+            "CHAINS_TEST2_BATCHCONTRACTADDRESS",
+            multicall_address_string.clone(),
+        )
+        .hyp_env("CHAINS_TEST2_MAXBATCHSIZE", "5")
+        .hyp_env("CHAINS_TEST3_CONNECTION_URL", "http://127.0.0.1:8545")
+        .hyp_env(
+            "CHAINS_TEST3_BATCHCONTRACTADDRESS",
+            multicall_address_string,
+        )
+        .hyp_env("CHAINS_TEST3_MAXBATCHSIZE", "5")
+        .hyp_env("METRICSPORT", RELAYER_METRICS_PORT)
+        .hyp_env("DB", relayer_db.to_str().unwrap())
+        .hyp_env("CHAINS_TEST1_SIGNER_KEY", RELAYER_KEYS[0])
+        .hyp_env("CHAINS_TEST2_SIGNER_KEY", RELAYER_KEYS[1])
+        .hyp_env("CHAINS_SEALEVELTEST1_SIGNER_KEY", RELAYER_KEYS[3])
+        .hyp_env("CHAINS_SEALEVELTEST2_SIGNER_KEY", RELAYER_KEYS[4])
+        .hyp_env("RELAYCHAINS", "invalidchain,otherinvalid")
+        .hyp_env("ALLOWLOCALCHECKPOINTSYNCERS", "true")
+        .hyp_env(
+            "GASPAYMENTENFORCEMENT",
+            r#"[{
+                "type": "minimum",
+                "payment": "1"
+            }]"#,
+        )
+        .arg(
+            "chains.test1.customRpcUrls",
+            "http://127.0.0.1:8545,http://127.0.0.1:8545,http://127.0.0.1:8545",
+        )
+        // default is used for TEST3
+        .arg("defaultSigner.key", RELAYER_KEYS[2]);
+    if config.sealevel_enabled {
+        relayer_env.arg(
+            "relayChains",
+            "test1,test2,test3,sealeveltest1,sealeveltest2",
+        )
+    } else {
+        relayer_env.arg("relayChains", "test1,test2,test3")
+    }
+}
+
+fn restart_relayer(config: &Config, state: &mut State, rocks_db_dir: &TempDir) {
+    let (child, _) = state.agents.get_mut("RLY").expect("No relayer agent found");
+
+    child.kill().expect("Failed to kill relayer");
+
+    let relayer_env = create_relayer(config, rocks_db_dir);
+
+    log!("Restarting relayer...");
+    state.push_agent(relayer_env.spawn("RLY", Some(&AGENT_LOGGING_DIR)));
+}
+
+fn message_retrieval_invariants_met() -> eyre::Result<bool> {
+    let log_file_path = AGENT_LOGGING_DIR.join("RLY-output.log");
+    let relayer_logfile = File::open(log_file_path).unwrap();
+
+    log!("Checking message statuses were retrieved from logs...");
+
+    const RETRIEVED_MESSGE_LOG: &str = "Message status retrieved from db";
```

Comment:
if we use this in e2e can you put this const in the relayer and we import from it? that way we're less likely to accidentally break e2e if we change a log line. iirc we can import from the relayer crate in here

Reply:
also nit typo `RETRIEVED_MESSGE_LOG` -> `RETRIEVED_MESSAGE_LOG`